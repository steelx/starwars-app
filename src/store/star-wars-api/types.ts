export interface StarWarsAPIResponse {
    count: number;
    next: string | null;
    previous: string | null;
    results: Character[];
}

export interface Film {
    url: string; // Assuming you don't need details about films
}

export interface Character {
    name: string;
    height: string;
    mass: string;
    hair_color: string;
    skin_color: string;
    eye_color: string;
    birth_year: string;
    gender: string;
    homeworld: string;
    films: Film[]; // Array of film URLs
    species: string[]; // Assuming you don't need details about species
    vehicles: string[]; // Array of vehicle URLs
    starships: string[]; // Array of starship URLs
    created: string;
    edited: string;
    url: string;
}
