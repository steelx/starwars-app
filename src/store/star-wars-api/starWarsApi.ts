import {createApi, fetchBaseQuery} from "@reduxjs/toolkit/query/react";
import {Character, StarWarsAPIResponse} from "@store/star-wars-api/types";
import {QueryArgs} from "@store/types";

const BASE_URL = 'https://swapi.dev/api';
export const starWarsApi = createApi({
    reducerPath: 'starWarsApi',
    baseQuery: fetchBaseQuery({
        baseUrl: BASE_URL
    }),
    endpoints: (build) => ({
        getCharacters: build.query<Character[], QueryArgs>({
            query: (queryArgs) => ({
                url: `/people?page=${queryArgs.page ?? 1}`,
            }),
            transformResponse: (response: StarWarsAPIResponse) => response.results
        })
    })
});

export const { useGetCharactersQuery } = starWarsApi
