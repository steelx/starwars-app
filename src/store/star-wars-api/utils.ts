export const getFilmIds = (filmUrls: string[]): string[] | null => {
    if (!filmUrls.length) return null;

    const idRegex = /\/api\/films\/(\d+)\//;
    const filmIds = filmUrls.map(url => {
        const match = url.match(idRegex);
        return match ? match[1] : null;
    })

    return filmIds.every(id => typeof id === 'string') ? filmIds as string[] : null;
}
