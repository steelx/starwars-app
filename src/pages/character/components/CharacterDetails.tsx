import React from "react";
import {ICharacter} from "@store/star-wars-api/types";
import {
    CharacterDetailsWrapper,
    CharacterDetailsContainer,
    FilmList
} from "@pages/character/components/styles";
import {useGetCharacterFilmsQuery} from "@store/star-wars-api/starWarsApi";
import CharacterDetailsItem from "@pages/character/components/CharacterDetailsItem";


interface CharacterProps {
    character: ICharacter
}

const CharacterDetails: React.FC<CharacterProps> = ({ character }) => {
    const { data: films, isLoading: isFilmDetailsLoading } =
        useGetCharacterFilmsQuery(character.films)

    return <CharacterDetailsWrapper>
        <CharacterDetailsContainer>
            <h2>{character.name}</h2>
            <ul>
                <CharacterDetailsItem>Hair Color: {character.hair_color}</CharacterDetailsItem>
                <CharacterDetailsItem>Eye Color: {character.eye_color}</CharacterDetailsItem>
                <CharacterDetailsItem>Height: {character.height}</CharacterDetailsItem>
                <CharacterDetailsItem>Gender: {character.gender}</CharacterDetailsItem>
                <CharacterDetailsItem>Homeworld: {character.homeworld}</CharacterDetailsItem>
            </ul>
        </CharacterDetailsContainer>
        <FilmList>
            <h3>Films Appeared In:</h3>
            <ul>
                { films && films.length > 0 && (
                    films.map((detail, index) => (
                        <CharacterDetailsItem key={index}>{detail.title}</CharacterDetailsItem>
                    ))
                )}
                { isFilmDetailsLoading && (
                    <CharacterDetailsItem>Fetching film titles...</CharacterDetailsItem>
                )}
            </ul>
        </FilmList>
    </CharacterDetailsWrapper>
}

export default CharacterDetails;
