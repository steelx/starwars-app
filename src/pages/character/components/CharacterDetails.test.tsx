import {MockedFunction} from "vitest";
import {render} from "@testing-library/react";
import CharacterDetails from "@pages/character/components/CharacterDetails";
import {useGetCharacterFilmsQuery} from "@store/star-wars-api/starWarsApi";
import {mockCharacter} from "@test-helpers/character-data";

vi.mock("@store/star-wars-api/starWarsApi");

describe("CharacterDetails", () => {
    const mockUseGetCharacterFilmsQuery = useGetCharacterFilmsQuery as MockedFunction<typeof useGetCharacterFilmsQuery>;

    afterEach(() => {
        vi.resetAllMocks()
    })

    test("should render character details", async () => {
        mockUseGetCharacterFilmsQuery.mockReturnValueOnce({
            isLoading: true,
            refetch: vi.fn()
        })

        const {asFragment} = render(<CharacterDetails character={mockCharacter} />)
        expect(asFragment()).toMatchSnapshot()
    })

    test("should display `Fetching film titles...` while films is loading", async () => {
        mockUseGetCharacterFilmsQuery.mockReturnValueOnce({
            isLoading: true,
            refetch: vi.fn()
        })

        const { getByText } = render(<CharacterDetails character={mockCharacter} />)
        expect(getByText("Fetching film titles...")).toBeTruthy()
    })

    test("should display films appeared in", async () => {
        mockUseGetCharacterFilmsQuery.mockReturnValueOnce({
            data: [
                { title: "Lalola" },
                { title: "Star wars movie" },
            ],
            isLoading: false,
            refetch: vi.fn()
        })

        const { getByText } = render(<CharacterDetails character={mockCharacter} />)
        expect(getByText("Lalola")).toBeTruthy()
        expect(getByText("Star wars movie")).toBeTruthy()
    })
});
