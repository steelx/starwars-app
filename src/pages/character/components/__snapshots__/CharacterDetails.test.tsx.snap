// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CharacterDetails > should render character details 1`] = `
<DocumentFragment>
  <div
    class="sc-beySPh fOSsff"
  >
    <div
      class="sc-guDLey jmmUSo"
    >
      <h2>
        C-3PO
      </h2>
      <ul>
        <li
          class="sc-dmyCSP cRcTuR"
        >
          Hair Color: n/a
        </li>
        <li
          class="sc-dmyCSP cRcTuR"
        >
          Eye Color: yellow
        </li>
        <li
          class="sc-dmyCSP cRcTuR"
        >
          Height: 167
        </li>
        <li
          class="sc-dmyCSP cRcTuR"
        >
          Gender: n/a
        </li>
        <li
          class="sc-dmyCSP cRcTuR"
        >
          Homeworld: https://swapi.dev/api/planets/1/
        </li>
      </ul>
    </div>
    <div
      class="sc-hLQSwg gDiYxq"
    >
      <h3>
        Films Appeared In:
      </h3>
      <ul>
        <li
          class="sc-dmyCSP cRcTuR"
        >
          Fetching film titles...
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
